src/libexpr/eval.cc

```diff
@@ -144,8 +144,11 @@ void printValue(std::ostream & str, std::set<const Value *> & active, const Valu
     case tFloat:
         str << v.fpoint;
         break;
+    case tBlackhole:
+        str << "<BLACKHOLE>";
+        break;
     default:
-        throw Error("invalid value");
+        throw Error("invalid value %1%", showType(v));
```

Comment:
I think that this aborts if the type is unknown.

Reply:
Oh yeah, you're right: should I replace the `abort` with a `throw` in there?

Reply:
Ok, now it doesn't abort: (intentional bug for testing)
```
ckie@cookiemonster ~/git/nixfiles -> nix eval --expr 'builtins.trace (x: 1) 1'
error: invalid type id 100
```

Reply:
No, I think we should continue to abort, since invalid values are bugs in Nix, not user errors.

Reply:
So actually I would change this line to `abort()` just as in `showType()`.

Reply:
That makes it significantly more painful to debug: I ran into the error message randomly so I grepped for it and found the source.
Can we log *and* abort?

Reply:
`abort()` makes it easier to debug since in gdb it puts you right on the spot where the crash happened. Whereas if you throw an exception, you have to mess around with `catch throw`.